File: LeetCodePractice/src/LinkedList/LRUcache.java
package LinkedList;

import java.util.*;
/*
Design and implement a data structure for Least Recently Used (LRU) cache. It should support the following operations: get and put.

        get(key) - Get the value (will always be positive) of the key if the key exists in the cache, otherwise return -1.
        put(key, value) - Set or insert the value if the key is not already present. When the cache reached its capacity, it should invalidate the least recently used item before inserting a new item.

        The cache is initialized with a positive capacity.

        Follow up:
        Could you do both operations in O(1) time complexity?

        Example:

        LRUCache cache = new LRUCache( 2 *//* capacity *//* );

        cache.put(1, 1);
        cache.put(2, 2);
        cache.get(1);       // returns 1
        cache.put(3, 3);    // evicts key 2
        cache.get(2);       // returns -1 (not found)
        cache.put(4, 4);    // evicts key 1
        cache.get(1);       // returns -1 (not found)
        cache.get(3);       // returns 3
        cache.get(4);       // returns 4

        solution:
        create Double LinkList Node (k,v)
        remove head.next when reach capacity
        move the node to tail.pre for get operation
        create private method insertBeforeTail
*/
public class LRUcache {
	private class Node {
        Node pre;
        Node next;
        int val;
        int key;
        public Node(int k, int v){
            key = k;
            val = v;
        }
    }
    private int capacity;
    private HashMap<Integer, Node> map = new HashMap<>();
    // point to the least recent one
    private Node head = new Node(-1, -1);
    // point to the most recent one
    private Node tail = new Node(-1, -1);

    // @param capacity, an integer
    public LRUcache(int capacity) {
        head.next = tail;
        tail.pre = head;
        this.capacity = capacity;
    }

    private void insertBeforeTail(Node n){
        tail.pre.next = n;
        n.next = tail;
        n.pre = tail.pre;
        tail.pre = n;
    }
    
    // @return an integer
    public int get(int key) {
        if(!map.containsKey(key)){
            return -1;
        }
        Node n = map.get(key);
        n.pre.next = n.next;
        n.next.pre = n.pre;
        insertBeforeTail(n);
        return n.val;
    }

    // @param key, an integer
    // @param value, an integer
    // @return nothing
    public void set(int key, int value) {
        // write your code here
        if(get(key) != -1){
            //exist node
            Node n = map.get(key);
            n.val = value;
        } else {
            // not exist node
            if(map.size() == capacity){
                // map full, need to remove least recent one
                map.remove(head.next.key);
                head.next = head.next.next;
                head.next.pre = head;
            }
            //add new node to map
            Node n = new Node(key, value);
            map.put(key, n);
            //add to latest
            insertBeforeTail(n);
        }
    }
}

File: LeetCodePython/LRUCache.py
"""
Design a data structure that follows the constraints of a Least Recently Used (LRU) cache.

Implement the LRUCache class:

LRUCache(int capacity) Initialize the LRU cache with positive size capacity.
int get(int key) Return the value of the key if the key exists, otherwise return -1.
void put(int key, int value) Update the value of the key if the key exists. Otherwise, add the key-value pair to the cache. If the number of keys exceeds the capacity from this operation, evict the least recently used key.
The functions get and put must each run in O(1) average time complexity.



Example 1:

Input
["LRUCache", "put", "put", "get", "put", "get", "put", "get", "get", "get"]
[[2], [1, 1], [2, 2], [1], [3, 3], [2], [4, 4], [1], [3], [4]]
Output
[null, null, null, 1, null, -1, null, -1, 3, 4]

Explanation
LRUCache lRUCache = new LRUCache(2);
lRUCache.put(1, 1); // cache is {1=1}
lRUCache.put(2, 2); // cache is {1=1, 2=2}
lRUCache.get(1);    // return 1
lRUCache.put(3, 3); // LRU key was 2, evicts key 2, cache is {1=1, 3=3}
lRUCache.get(2);    // returns -1 (not found)
lRUCache.put(4, 4); // LRU key was 1, evicts key 1, cache is {4=4, 3=3}
lRUCache.get(1);    // return -1 (not found)
lRUCache.get(3);    // return 3
lRUCache.get(4);    // return 4


Constraints:

1 <= capacity <= 3000
0 <= key <= 10^4
0 <= value <= 10^5
At most 2 * 10^5 calls will be made to get and put.
"""


class Node:
    def __init__(self, k, v):
        self.key = k
        self.val = v
        self.pre = None
        self.next = None

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.map = {}
        self.head = Node(-1, -1)
        self.tail = Node(-1, -1)
        self.head.next = self.tail
        self.tail.pre = self.head

    def get(self, key: int) -> int:
        if key not in self.map:
            return -1
        n = self.map[key]
        n.pre.next = n.next
        n.next.pre = n.pre
        self._insert_before_tail(n)
        return n.val

    def put(self, key: int, value: int) -> None:
        if key in self.map:
            n = self.map[key]
            n.val = value
            n.pre.next = n.next
            n.next.pre = n.pre
            self._insert_before_tail(n)
        else:
            if len(self.map) == self.capacity:
                self.map.pop(self.head.next.key)
                self.head.next = self.head.next.next
                self.head.next.pre = self.head
            n = Node(key, value)
            self.map[key] = n
            self._insert_before_tail(n)

    def _insert_before_tail(self, n: Node) -> None:
        self.tail.pre.next = n
        n.next = self.tail
        n.pre = self.tail.pre
        self.tail.pre = n


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)